influxql/ast.go

```diff
@@ -765,6 +767,32 @@ func (s *SelectStatement) RewriteWildcards(fields Fields, dimensions Dimensions)
 	return other
 }
 
+// RewriteDistinct rewrites the expresion to be a call for map/reduce to work correctly
+// This method assumes all validation has passed
+func (s *SelectStatement) RewriteDistinct() *SelectStatement {
+	if !s.HasDistinct() {
+		return s
+	}
+
+	other := s.Clone()
+
+	// Rewrite any `distinct foo` as `distinct(foo)`
+	rwFields := make(Fields, 0, len(s.Fields))
+	for _, f := range s.Fields {
+		switch v := f.Expr.(type) {
+		case *Distinct:
+			field := &Field{Expr: &Call{Name: "distinct", Args: []Expr{&VarRef{Val: v.Val}}}}
+			rwFields = append(rwFields, field)
+		default:
+			rwFields = append(rwFields, f)
+		}
+	}
+	other.Fields = rwFields
+	other.IsRawQuery = false
+
+	return other
+}
+
```

Comment:
I would simplify this method to look like this... https://gist.github.com/dgnorton/3125e459045a5cedc948.